File: quant1x/contrib/data/tdx/instruments.cpp
// Copyright (c) Quant1X <wangfengxy@sina.cn>.
// Licensed under the MIT License.
//
// tdx/instruments — TDX 证券信息缓存读取 + 初始化
// 与 Python contrib/data/tdx/instruments.py 和 Rust instruments.rs 对齐

#include "instruments.h"
#include <quant1x/config/base.h>
#include <quant1x/data/market.h>
#include <quant1x/data/meta/calendar.h>
#include <quant1x/data/status.h>
#include <quant1x/contrib/data/tdx/client.h>
#include <quant1x/contrib/data/tdx/protocol.h>
#include <quant1x/contrib/data/tdx/level1/security_list.h>
#include <quant1x/io/csv-reader.h>
#include <quant1x/io/csv-writer.h>
#include <quant1x/runtime/once.h>
#include <mutex>
#include <tsl/robin_map.h>
#include <algorithm>
#include <filesystem>

namespace tdx {
namespace instruments {

// ============================================================
// 常量
// ============================================================
constexpr int SECURITY_LIST_PRE_REQUEST_MAX = 1600;  ///< 单次请求最大证券数量

// ============================================================
// Exchange → TDX 市场 ID 映射
// 对齐 Python _EXCHANGE_TO_MARKET / Rust TDX_MARKET_* 常量
// ============================================================
inline int exchange_to_tdx_market(meta::Exchange ex) {
    switch (ex) {
        case meta::Exchange::SZSE: return 0;   // 深圳
        case meta::Exchange::SSE:  return 1;   // 上海
        case meta::Exchange::BSE:  return 2;   // 北京
        case meta::Exchange::HKEX: return 31;  // 香港 (扩展行情)
        case meta::Exchange::HKFE: return 27;  // 香港期货 (扩展行情)
        case meta::Exchange::USA:  return 74;  // 美国 (扩展行情)
        default:                   return -1;  // 不支持
    }
}

// ============================================================
// 内存缓存: symbol → Instrument
// RollingOnce 保证每日首次调用时初始化, 长期运行每天自动重新加载
// ============================================================
static auto security_once = RollingOnce::create("tdx-instruments", meta::cron_expr_daily_9am);
static std::mutex g_security_mutex;
static tsl::robin_map<std::string, meta::Instrument> g_security_map;

// ============================================================
// load_securities() — 从 CSV 加载到内存 (调用方负责通过 RollingOnce 控制时机)
// ============================================================
bool load_securities() {
    std::lock_guard<std::mutex> lock(g_security_mutex);

    std::string fname = config::get_meta_path() + "/securities.csv";
    spdlog::debug("[tdx/instruments] Loading securities from {}", fname);

    g_security_map.clear();

    try {
        io::CSVReader<9> in(fname);
        in.read_header(io::ignore_extra_column, "exchange", "type", "code", "name",
                       "lot_size", "price_precision", "ext_market", "ext_category", "alias_ticker");
        std::string exchange_str, type_str, ticker, name;
        std::string lot_str, prec_str, extm_str, extc_str, alias;
        int count = 0;
        while (in.read_row(exchange_str, type_str, ticker, name, lot_str, prec_str, extm_str, extc_str, alias)) {
            try {
                meta::Instrument inst;
                inst.exchange = meta::exchange_from_abbr(exchange_str);
                inst.type = meta::instype_from_string(type_str);
                for (auto& c : ticker) { c = static_cast<char>(::tolower(static_cast<unsigned char>(c))); }
                inst.ticker = ticker;
                inst.name = name;
                inst.lot_size = lot_str.empty() ? 100 : std::stoi(lot_str);
                inst.price_precision = prec_str.empty() ? 2 : std::stoi(prec_str);
                inst.ext_market = extm_str.empty() ? 0 : std::stoi(extm_str);
                inst.ext_category = extc_str.empty() ? 0 : std::stoi(extc_str);
                inst.alias_ticker = alias;

                std::string symbol = inst.symbol();
                g_security_map[symbol] = inst;
                count++;
            } catch (const std::exception& e) {
                spdlog::debug("[tdx/instruments] skip row: {} {} {} — {}", exchange_str, type_str, ticker, e.what());
            }
        }
        spdlog::info("[tdx/instruments] loaded {} instruments from {}", count, fname);
        return count > 0;
    } catch (const std::exception& e) {
        spdlog::warn("[tdx/instruments] cannot open or parse {}: {}", fname, e.what());
        return false;
    }
}

// ============================================================
// fetch_security_list() — 从 TDX 标准行情服务器获取一页证券
// 对齐 Python instruments.fetch_security_list() (标准行情部分)
// ============================================================
static std::vector<meta::Instrument> fetch_security_list(meta::Exchange exchange, int start, int count) {
    std::vector<meta::Instrument> result;
    int market_id = exchange_to_tdx_market(exchange);
    if (market_id < 0) {
        spdlog::error("[tdx/instruments] unsupported exchange for std market: {} (code={})",
                      meta::exchange_label(exchange), meta::exchange_code(exchange));
        return result;
    }

    try {
        auto conn = level1::get_std_conn();
        auto& sock = conn->socket();

        level1::SecurityList msg(market_id, start, count);
        auto err = level1::process(sock, msg);
        if (err) {
            spdlog::error("[tdx/instruments] SecurityList fetch failed: {} {}",
                          meta::exchange_code(exchange), err.message());
            return result;
        }

        spdlog::debug("[tdx/instruments] SecurityList {} start={} count={} got {} records",
                      meta::exchange_code(exchange), start, count, msg.Count);

        for (const auto& sec : msg.List) {
            meta::Instrument inst;
            inst.exchange = exchange;
            inst.type = meta::InstrumentType::Stock; // A 股默认 Stock 类型
            // 复制代码并转小写
            std::string code_lower = sec.Code;
            for (auto& c : code_lower) { c = static_cast<char>(::tolower(static_cast<unsigned char>(c))); }
            inst.ticker = code_lower;
            inst.name = sec.Name;
            inst.lot_size = static_cast<int>(sec.VolUnit);
            inst.price_precision = static_cast<int>(sec.DecimalPoint);
            inst.ext_market = market_id;      // 与 Python 对齐: ext_market = helpers.exchange_to_market(exchange)
            inst.ext_category = 0;

            result.push_back(inst);
        }

    } catch (const std::exception& e) {
        spdlog::error("[tdx/instruments] fetch_security_list exception: {}", e.what());
    }

    return result;
}

// ============================================================
// write_securities_csv() — 将证券列表写入 CSV
// 对齐 Rust write_securities_csv()
// ============================================================
static void write_securities_csv(const std::string& fname, const std::vector<meta::Instrument>& instruments) {
    // Ensure parent directory exists
    std::filesystem::path p(fname);
    if (p.has_parent_path()) {
        std::filesystem::create_directories(p.parent_path());
    }

    try {
        io::CSVWriter writer(fname);
        writer.write_row("exchange", "type", "code", "name", "lot_size", "price_precision",
                         "ext_market", "ext_category", "alias_ticker");

        for (const auto& inst : instruments) {
            writer.write_row(
                meta::exchange_identifier(inst.exchange),
                meta::instype_to_string(inst.type),
                inst.ticker,
                inst.name,
                inst.lot_size,
                inst.price_precision,
                inst.ext_market,
                inst.ext_category,
                inst.alias_ticker
            );
        }
        spdlog::info("[tdx/instruments] wrote {} instruments to {}", instruments.size(), fname);
    } catch (const std::exception& e) {
        spdlog::error("[tdx/instruments] cannot create {}: {}", fname, e.what());
    }
}

// ============================================================
// do_init_securities() — 实际初始化逻辑, 由 RollingOnce::Do 调用
// ============================================================
static void do_init_securities() {
    std::string fname = config::get_meta_path() + "/securities.csv";

    // Step 1: 检查是否需要更新
    bool create_or_update = data::should_initialize_file(fname);
    if (!create_or_update) {
        // CSV 存在且是今天的, 尝试加载
        create_or_update = !load_securities();
    }
    spdlog::debug("[tdx/instruments] init_securities create_or_update={}", create_or_update);

    if (!create_or_update) {
        return; // 已加载, 无需更新
    }

    // Step 2: 从 TDX 服务器拉取
    std::vector<meta::Instrument> instruments;

    // 2a. 标准行情: A 股 (SSE/SZSE/BSE)
    // 对齐 Python: markets = [Exchange.SSE, Exchange.SZSE, Exchange.BSE]
    std::vector<meta::Exchange> std_markets = {
        meta::Exchange::SSE,
        meta::Exchange::SZSE,
        meta::Exchange::BSE,
    };

    for (auto m : std_markets) {
        int start = 0;
        std::vector<meta::Instrument> rows;
        while (true) {
            auto page = fetch_security_list(m, start, SECURITY_LIST_PRE_REQUEST_MAX);
            if (page.empty()) {
                break;
            }
            size_t page_size = page.size();
            rows.insert(rows.end(), page.begin(), page.end());
            if (page_size < static_cast<size_t>(SECURITY_LIST_PRE_REQUEST_MAX)) {
                break;
            }
            start += SECURITY_LIST_PRE_REQUEST_MAX;
        }

        // 相同市场按代码排序 (对齐 Python: rows.sort(key=lambda x: x.ticker))
        std::sort(rows.begin(), rows.end(),
                  [](const meta::Instrument& a, const meta::Instrument& b) {
                      return a.ticker < b.ticker;
                  });

        spdlog::info("[tdx/instruments] fetched {} instruments from {}", rows.size(),
                     meta::exchange_label(m));
        instruments.insert(instruments.end(), rows.begin(), rows.end());
    }

    // 2b. TODO: 扩展行情 (HKEX 等)
    // 需要 ExtensionProtocolHandler + ExtensionConnectionPool + InstrumentInfo 消息类
    // 等 ext 协议基础设施完成后接入
    // 对齐 Python instruments.py init_securities() 第 127-158 行
    if (true) {
        // 以下为预留框架, 待 ext 协议层完成后实现:
        // std::vector<meta::Exchange> ext_markets = { meta::Exchange::HKEX };
        // for (auto m : ext_markets) { ... }
        spdlog::info("[tdx/instruments] ext market (HKEX) not yet implemented — requires ext protocol handler");
    }

    // Step 3: 写入 CSV
    if (!instruments.empty()) {
        write_securities_csv(fname, instruments);
    } else {
        spdlog::warn("[tdx/instruments] no instruments fetched — CSV not written");
    }

    // Step 4: 加载到内存 (load_securities 内部会加锁并覆盖 g_security_map)
    bool ok = load_securities();
    if (!ok) {
        spdlog::error("[tdx/instruments] failed to load securities after initialization");
    }
}

// ============================================================
// init_securities() — 通过 RollingOnce 保证每日首次调用时初始化
// 对齐 Python init_securities() / Rust init_securities()
// ============================================================
void init_securities() {
    security_once->Do(do_init_securities);
}

// ============================================================
// get_code_list() — 返回所有 symbol 字符串
// ============================================================
std::vector<std::string> get_code_list() {
    init_securities(); // RollingOnce 保证每天只执行一次

    std::lock_guard<std::mutex> lock(g_security_mutex);
    std::vector<std::string> codes;
    codes.reserve(g_security_map.size());
    for (const auto& [symbol, _] : g_security_map) {
        codes.push_back(symbol);
    }
    return codes;
}

// ============================================================
// get_instrument_info() — 查找单个证券
// ============================================================
std::optional<meta::Instrument> get_instrument_info(const std::string& symbol) {
    std::string security_code = data::correct_security_code(symbol);
    spdlog::debug("[tdx/instruments] get_instrument_info: symbol={}, security_code={}", symbol, security_code);

    init_securities(); // RollingOnce 保证每天只执行一次

    std::lock_guard<std::mutex> lock(g_security_mutex);
    auto it = g_security_map.find(security_code);
    if (it != g_security_map.end()) {
        return it->second;
    }
    return std::nullopt;
}

// ============================================================
// EnsureSecuritiesInitialized() — 供外部在策略启动时调用
// ============================================================
void EnsureSecuritiesInitialized() {
    init_securities();
}

} // namespace instruments
} // namespace tdx

File: quant1x/contrib/data/tdx/instruments.go
// Copyright (c) Quant1X <wangfengxy@sina.cn>.
// Licensed under the MIT License.
//
// tdx/instruments - TDX 证券列表缓存加载与初始化
// 对齐 Python contrib/data/tdx/instruments.py 与 Rust instruments.rs / C++ instruments.cpp

package tdx

import (
	"encoding/csv"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"sync"

	"github.com/quant1x/quant1x/quant1x/config"
	"github.com/quant1x/quant1x/quant1x/contrib/data/tdx/level1/std"
	"github.com/quant1x/quant1x/quant1x/contrib/data/tdx/tdxproto"
	"github.com/quant1x/quant1x/quant1x/data"
	"github.com/quant1x/quant1x/quant1x/data/meta"
	"github.com/quant1x/quant1x/quant1x/runtime"
	logger "github.com/quant1x/quant1x/quant1x/log"
)

// SecurityListPerRequestMax 单次请求最大证券数量, 与 Python/C++/Rust 对齐
const SecurityListPerRequestMax = 1600

// securityMap 内存缓存: symbol -> Instrument
var (
	securityOnce  = runtime.RollingOnceFromSpec(config.GLOBAL_CRON_EXPR_DAILY_INIT)
	securityMutex sync.RWMutex
	securityMap   = make(map[string]meta.Instrument)
)

// loadSecurities 从 CSV 加载到内存 (调用方负责通过 RollingOnce 控制时机)
func loadSecurities() bool {
	fname := config.GetSecurityFilename()
	logger.Debugf("[tdx/instruments] Loading securities from %s", fname)

	securityMutex.Lock()
	defer securityMutex.Unlock()

	securityMap = make(map[string]meta.Instrument)

	file, err := os.Open(fname)
	if err != nil {
		return false
	}
	defer file.Close()

	reader := csv.NewReader(file)
	headers, err := reader.Read()
	if err != nil {
		return false
	}
	_ = headers

	count := 0
	for {
		record, err := reader.Read()
		if err != nil {
			break
		}
		if len(record) < 8 {
			continue
		}
		exchangeStr := record[0]
		if exchangeStr == "" {
			exchangeStr = "unknown"
		}
		exchange, err := meta.ParseExchange(exchangeStr)
		if err != nil {
			exchange = meta.UNKNOWN
		}
		typeStr := record[1]
		if typeStr == "" {
			typeStr = "unknown"
		}
		instType := meta.InstrumentTypeFromString(typeStr)
		ticker := strings.ToLower(record[2])
		name := record[3]
		lotSize := parseInt(record[4], 100)
		pricePrecision := parseInt(record[5], 2)
		extMarket := parseInt(record[6], 0)
		extCategory := parseInt(record[7], 0)

		inst := meta.Instrument{
			Exchange:       exchange,
			Type:           instType,
			Ticker:         ticker,
			Name:           name,
			LotSize:        lotSize,
			PricePrecision: pricePrecision,
			ExtMarket:      extMarket,
			ExtCategory:    extCategory,
		}
		symbol := inst.Symbol()
		securityMap[symbol] = inst
		count++
	}
	logger.Infof("[tdx/instruments] loaded %d instruments from %s", count, fname)
	return count > 0
}

// parseInt 解析整数字符串, 失败时返回默认值
func parseInt(s string, def int) int {
	s = strings.TrimSpace(s)
	if s == "" {
		return def
	}
	v, err := strconv.Atoi(s)
	if err != nil {
		return def
	}
	return v
}

// fetchSecurityList 从 TDX 标准行情服务器获取一页证券
// 对齐 Python instruments.fetch_security_list() (标准行情部分)
func fetchSecurityList(exchange meta.Exchange, start, count int) []meta.Instrument {
	var result []meta.Instrument
	if tdxproto.ExchangeToMarketId(exchange) < 0 {
		logger.Errorf("[tdx/instruments] unsupported exchange for std market: %s", exchange)
		return result
	}

	conn, release, err := GetStdConnection()
	if err != nil {
		logger.Errorf("[tdx/instruments] get std connection failed: %v", err)
		return result
	}
	if release != nil {
		defer release()
	}
	if conn == nil || conn.Conn() == nil {
		logger.Errorf("[tdx/instruments] nil connection from level1 client")
		return result
	}

	msg := std.NewSecurityListContext(exchange, start, count)
	if err := tdxproto.TransactMessageSync(conn, msg); err != nil {
		logger.Errorf("[tdx/instruments] SecurityListContext fetch failed: %s %v", exchange, err)
		return result
	}

	logger.Debugf("[tdx/instruments] SecurityListContext %s start=%d count=%d got %d records",
		exchange, start, count, len(msg.List))

	for _, sec := range msg.List {
		inst := meta.Instrument{
			Exchange:       exchange,
			Ticker:         strings.ToLower(sec.Code),
			Name:           sec.Name,
			LotSize:        int(sec.VolUnit),
			PricePrecision: int(sec.DecimalPoint),
			ExtMarket:      tdxproto.ExchangeToMarketId(exchange), // 与 Python 对齐: ext_market = exchange_to_market(exchange)
		}
		inst.Type = data.DetectInstrumentTypeByRule(exchange, inst.Ticker)
		inst.ExtCategory = int(inst.Type)
		result = append(result, inst)
	}
	return result
}

// writeSecuritiesCsv 将证券列表写入 CSV
func writeSecuritiesCsv(fname string, instruments []meta.Instrument) {
	// Ensure parent directory exists
	if dir := filepath.Dir(fname); dir != "" {
		_ = os.MkdirAll(dir, 0o755)
	}

	file, err := os.Create(fname)
	if err != nil {
		logger.Errorf("[tdx/instruments] cannot create %s: %v", fname, err)
		return
	}
	defer file.Close()

	writer := csv.NewWriter(file)
	defer writer.Flush()

	var zeroInstrument meta.Instrument
	headers := zeroInstrument.Headers()
	if err := writer.Write(headers); err != nil {
		logger.Errorf("[tdx/instruments] cannot write header: %v", err)
		return
	}

	for i := range instruments {
		row := instruments[i].ToSlice()
		strRow := make([]string, len(row))
		for j, cell := range row {
			strRow[j] = cellToString(cell)
		}
		if err := writer.Write(strRow); err != nil {
			logger.Errorf("[tdx/instruments] cannot write row: %v", err)
			return
		}
	}
	writer.Flush()
	if err := writer.Error(); err != nil {
		logger.Errorf("[tdx/instruments] cannot flush %s: %v", fname, err)
		return
	}
	logger.Infof("[tdx/instruments] wrote %d instruments to %s", len(instruments), fname)
}

// cellToString 将 CSV 单元格值转为字符串
func cellToString(v any) string {
	switch t := v.(type) {
	case string:
		return t
	case meta.InstrumentType:
		return t.String()
	case int:
		return strconv.Itoa(t)
	default:
		return ""
	}
}

// doInitSecurities 实际初始化逻辑, 由 RollingOnce::Do 调用
func doInitSecurities() {
	fname := config.GetSecurityFilename()

	// Step 1: 检查是否需要更新
	createOrUpdate := meta.ShouldInitializeFile(fname)
	if !createOrUpdate {
		// CSV 存在且是今天的, 尝试加载
		createOrUpdate = !loadSecurities()
	}
	logger.Debugf("[tdx/instruments] init_securities create_or_update=%v", createOrUpdate)

	if !createOrUpdate {
		return // 已加载, 无需更新
	}

	// Step 2: 从 TDX 服务器拉取
	var instruments []meta.Instrument

	// 2a. 标准行情: A 股 (SSE/SZSE/BSE)
	// 对齐 Python: markets = [Exchange.SSE, Exchange.SZSE, Exchange.BSE]
	stdMarkets := []meta.Exchange{meta.SSE, meta.SZSE, meta.BSE}
	for _, m := range stdMarkets {
		start := 0
		var rows []meta.Instrument
		for {
			page := fetchSecurityList(m, start, SecurityListPerRequestMax)
			if len(page) == 0 {
				break
			}
			pageSize := len(page)
			rows = append(rows, page...)
			if pageSize < SecurityListPerRequestMax {
				break
			}
			start += SecurityListPerRequestMax
		}

		// 相同市场按代码排序 (对齐 Python: rows.sort(key=lambda x: x.ticker))
		sort.Slice(rows, func(i, j int) bool {
			return rows[i].Ticker < rows[j].Ticker
		})

		logger.Infof("[tdx/instruments] fetched %d instruments from %s", len(rows), m)
		instruments = append(instruments, rows...)
	}

	// 2b. TODO: 扩展行情 (HKEX 等), 与 C++ 对齐, 等 ext 协议基础设施完成后接入
	logger.Infof("[tdx/instruments] ext market (HKEX) not yet implemented - requires ext protocol handler")

	// Step 3: 写入 CSV
	if len(instruments) > 0 {
		writeSecuritiesCsv(fname, instruments)
	} else {
		logger.Warnf("[tdx/instruments] no instruments fetched, CSV not written")
	}

	// Step 4: 加载到内存
	if !loadSecurities() {
		logger.Errorf("[tdx/instruments] failed to load securities after initialization")
	}
}

// InitSecurities 通过 RollingOnce 保证每日首次调用时初始化
// 对齐 Python init_securities() / Rust init_securities()
func InitSecurities() {
	securityOnce.Do(doInitSecurities)
}

// GetCodeList 返回所有 symbol 字符串
// 对齐 C++ get_code_list()
func GetCodeList() []string {
	InitSecurities() // RollingOnce 保证每天只执行一次
	securityMutex.RLock()
	defer securityMutex.RUnlock()

	codes := make([]string, 0, len(securityMap))
	for symbol := range securityMap {
		codes = append(codes, symbol)
	}
	return codes
}

// GetInstrumentInfo 查找单个证券
// 对齐 C++ get_instrument_info()
func GetInstrumentInfo(symbol string) *meta.Instrument {
	securityCode := data.CorrectSecurityCode(symbol)
	logger.Debugf("[tdx/instruments] get_instrument_info: symbol=%s, security_code=%s", symbol, securityCode)

	InitSecurities() // RollingOnce 保证每天只执行一次
	securityMutex.RLock()
	defer securityMutex.RUnlock()

	if inst, ok := securityMap[securityCode]; ok {
		return &inst
	}
	return nil
}

// EnsureSecuritiesInitialized 供外部在策略启动时调用
// 对齐 C++ ensure_securities_initialized()
func EnsureSecuritiesInitialized() {
	InitSecurities()
}
